src/assets/scripts/client/trafficGenerator/SpawnPatternModel.js

```diff
@@ -0,0 +1,487 @@
+import _forEach from 'lodash/forEach';
+import _map from 'lodash/map';
+import _isArray from 'lodash/isArray';
+import _isEmpty from 'lodash/isEmpty';
+import _isObject from 'lodash/isObject';
+import _random from 'lodash/random';
+import BaseModel from '../base/BaseModel';
+import RouteModel from '../airport/Route/RouteModel';
+import { bearingToPoint } from '../math/flightMath';
+import { AIRPORT_CONSTANTS } from '../constants/airportConstants';
+import { FLIGHT_CATEGORY } from '../constants/aircraftConstants';
+import { TIME } from '../constants/globalConstants';
+
+/**
+ * Defines a spawn pattern for a specific route within the area
+ *
+ * This can be for departures or arrivals. Provides a starting point and definition
+ * for all spawning aircraft
+ *
+ * @class SpawnPatternModel
+ * @extends BaseModel
+ */
+export default class SpawnPatternModel extends BaseModel {
+    /**
+     * @constructor
+     * @for SpawnPatternModel
+     * @param spawnPatternJson {object}
+     * @param navigationLibrary {NavigationLibrary}
+     */
+    /* istanbul ignore next */
+    constructor(spawnPatternJson, navigationLibrary) {
+        super(spawnPatternJson, navigationLibrary);
+
+        if (!_isObject(spawnPatternJson) || _isEmpty(spawnPatternJson)) {
+            throw new TypeError('Invalid parameter passed to SpawnPatternModel');
+        }
+
+        if (!_isObject(navigationLibrary) || _isEmpty(navigationLibrary)) {
+            throw new TypeError('Invalid NavigationLibrary passed to SpawnPatternModel');
+        }
+
+        /**
+         * Interval id
+         *
+         * Stored here so a specific interval can be associated with a
+         * specfic `SpawnPatternModel` instance. An Interval may be reset
+         * or changed during the life of the app.
+         *
+         * Provides easy access to a specif interval id
+         *
+         * @property scheduleId
+         * @type {number}
+         * @default -1
+         */
+        this.scheduleId = -1;
+
+        /**
+         * One of `FLIGHT_CATEGORY`
+         *
+         * @property category
+         * @type {string}
+         * @default ''
+         */
+        this.category = '';
+
+        /**
+         * Type of arrival or departure pattern
+         *
+         * Could be `random`, `cyclic`. `surge` or `wave`
+         *
+         * @property type
+         * @type {string}
+         * @default ''
+         */
+        this.method = '';
+
+        /**
+         *
+         *
+         * @property origin
+         * @type {string}
+         * @default ''
+         */
+        this.origin = '';
+
+        /**
+         *
+         *
+         * @property destination
+         * @type {string}
+         * @default ''
+         */
+        this.destination = '';
+
+        /**
+         * String representation of a `StandardRoute`
+         *
+         * @property route
+         * @type {string}
+         * @default
+         */
+        this.route = '';
```

Comment:
we could.  most of these class properties are either copied from the original `ArrivalBase` and/or coming from the `spawnPatterns` object received from the airport json.  I do think it would be helpful to add a `_routeModel` property and then add `route`/`routeString` or whatever as a getter.  But I'm only using the `RouteModel` in one spot, so that doesn't really make sense.

The reason I've avoided using `routeString` or had Abe refactor that to `routeCode` is that it is bad [programming] practice to name a var with a type (like`routeString`).  But if `routeString` is actually a technical term, then thats a different story.  That would make the argument for `routeString` much easier.  And also adds a little more need for that dictionary issue I added a while ago 😉